### PR TITLE
Add sanitizers

### PR DESCRIPTION
**Pull request by pfagner**
- AddressSanitizer: detects addressability issues;
- MemorySanitizer: detects use of uninitialized memory;
- ThreadSanitizer: detects data races and deadlocks;
- LeakSanitizer: detects memory leaks;
- UndefinedBehaviorSanitizer: detect undefined behavior issues.

### CODE REVIEW

File: tests/sanitizers/examples/asan_01.cc [hunk 1]
```diff
@@ -0,0 +1,5 @@
+int main(int argc, char **argv) {
+  int *array = new int[100];
+  delete [] array;
+  return array[argc];  // BOOM
+}
```
Comment (bitmouse) [inline tests/sanitizers/examples/asan_01.cc:5]: ?????

File: tests/sanitizers/examples/asan_02.cc [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#include <stdlib.h>
+#include <stdio.h>
+#define N 10
+int main(void) {
+  char s[N]  = "123456789";
+  for (int i = 0; i <= N; i++)   // Loop bound goes past the end of the array
+    printf ("%c", s[i]);
+  printf("\n");
+  return 0;
+}
```
Comment (bitmouse) [inline tests/sanitizers/examples/asan_02.cc:10]: ???????

File: tests/sanitizers/examples/asan_03.cc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#include <stdlib.h>
+#include <stdio.h>
+int main(void) {
+  char *s = (char*)malloc(sizeof(char) * 10);
+  free(s);
+  free(s);  // DOUBLE FREE
+  printf("%s", s);
+  return 0;
+}
```
Comment (bitmouse) [inline tests/sanitizers/examples/asan_03.cc:9]: Why you add these?

**Review by bitmouse — CHANGES_REQUESTED** [review]

File: tests/sanitizers/examples/asan_01.cc (hunk 1 above)
Comment (pfagner) [reply tests/sanitizers/examples/asan_01.cc:5]: This is test case for sanitizer

Comment (bitmouse) [reply tests/sanitizers/examples/asan_01.cc:5]: We don't test sanitizers. We use them.

File: tests/sanitizers/test_sanitizers.sh [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+# │  │   └─ test_sanitizers.sh*   <--- executable script
+
+function outerdir(){
+   echo "$1" | sed -E 's@/[^.]+.*@/@'
+}
+DIR_BUILD=$PWD
+DIR_OUTER=$(outerdir "$0")
+if [ "$DIR_OUTER" = "./" ]
+then
+     echo "The script must be run from the 'facelift/build'"
```
Comment (jacky309) [inline tests/sanitizers/test_sanitizers.sh:36]: we do not want that. See my remark on the other PR.

Comment (pfagner) [reply tests/sanitizers/test_sanitizers.sh:36]: Okay.

**Review by bitmouse — CHANGES_REQUESTED** [review]
I am unable to generate any useful output using this instruction. I can see only build output and this is not what I expect.
Please generate report from e.g. running tests. Suggestions review.

File: tests/sanitizers/test_sanitizers.sh [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+SANITIZER_LDFLAGS="${!SANITIZER}"
+SANITIZER_FLAGS="${!SANITIZER}${!FLAGS}"
+SANITIZER_FLAGS+="$EXTRA_FLAGS"
+SANITIZER_FLAGS+="-shared-libasan "
+# NOTE: option '-shared-libasan' for link shared library with:
+# ASAN as a shared library. See more: https://github.com/google/sanitizers/wiki/AddressSanitizerAsDso
+# UBSAN as a shared library. See more: https://reviews.llvm.org/D38525
+
+CXX_VERSION_OUTPUT=$(clang --version)
+CXX_VERSION=$(echo $CXX_VERSION_OUTPUT | cut -d"-" -f1 | cut -d" " -f4)
```
Comment (bitmouse) [inline tests/sanitizers/test_sanitizers.sh:43]: This prints nothing
$ echo $CXX_VERSION



$ echo $CXX_VERSION_OUTPUT
clang version 9.0.0-2 (tags/RELEASE_900/final) Target: x86_64-pc-linux-gnu Thread model: posix InstalledDir: /usr/bin


try 
clang --version | head -n 1 | grep -o -E "[[:digit:]].[[:digit:]].[[:digit:]]" | uniq | sort

File: tests/sanitizers/test_sanitizers.sh [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    ;;
+    *)
+       echo "Sanitizer not specified"
+    ;;
+esac
+
+# Run tests
+case "${SANITIZER}" in
+  "ASAN")
+    PATH_LIBCLANG_RT=$(locate $LIBCLANG_RT_ASAN | grep clang | grep $CXX_VERSION)
```
Comment (bitmouse) [inline tests/sanitizers/test_sanitizers.sh:80]: because $CXX_VERSION is empty in my case, this fails in this line.

**Review by bitmouse — CHANGES_REQUESTED** [review]

File: tests/sanitizers/test_sanitizers.sh (hunk 5 above)
Comment (pfagner) [reply tests/sanitizers/test_sanitizers.sh]: The expression contain regex that have a mistake, but it is not a problem.
True way: is read source code (#define ...), but not an interpretation output for human.

**Review by bitmouse — APPROVED** [review]

**Pull request merged**